[lldb][Mach-O corefiles] Don't init Target arch to corefile

jasonmolenda opened the pull request:
This patch is making three changes, when loading a Mach-O corefile:

1. At the start of `DoLoadCore`, if a binary was provided in addition to the corefile, initialize the Target's ArchSpec.

2. Before ProcessMachCore does its "exhaustive search" fallback, looking through the corefile contents for a userland dyld or mach kernel, we must make sure the Target has an ArchSpec, or methods that check the address word size, or initialize a DataExtractor based on the Target arch will not succeed.

3. Add logging when setting the Target's arch listing exactly what that setting was based on -- the corefile itself, or the main binary.

Jonas landed a change last August (started with a patch from me) which removed the Target ArchSpec initialization at the start of DoLoadCore, in a scenario where the corefile had arch armv7 and the main binary had arch armv7em (Cortex-M), and there was python code in the main binary's dSYM which sets the operating system threads provider based on the Target arch.  It did different things for armv7 or armv7em, and so it would fail.

Jonas' patch removed any ArchSpec setting at the start of DoLoadCore, so we wouldn't have an incorrect arch value, but that broke the exhaustive search for kernel binaries, because we didn't have an address word size or endianness.

This patch should navigate the needs of both use cases.

I spent a good bit of time trying to construct a test to capture all of these requirements -- but it turns out to be a good bit difficult, encompassing both a genuine kernel corefiles and a microcontroller firmware corefiles.

rdar://146821929

jasonmolenda committed:
[lldb][Mach-O corefiles] Don't init Target arch to corefile

This patch is making three changes, when loading a Mach-O
corefile:

1. At the start of `DoLoadCore`, if a binary was provided in addition
to the corefile, initialize the Target's ArchSpec.

2. Before ProcessMachCore does its "exhaustive search" fallback,
looking through the corefile contents for a userland dyld or mach
kernel, we must make sure the Target has an ArchSpec, or methods
that check the address word size, or initialize a DataExtractor
based on the Target arch will not succeed.

3. Add logging when setting the Target's arch listing exactly what
that setting was based on -- the corefile itself, or the main binary.

Jonas landed a change last August (started with a patch from me)
which removed the Target ArchSpec initialization at the start of
DoLoadCore, in a scenario where the corefile had arch armv7 and
the main binary had arch armv7em (Cortex-M), and there was python
code in the main binary's dSYM which sets the operating system threads
provider based on the Target arch.  It did different things for
armv7 or armv7em, and so it would fail.

Jonas' patch removed any ArchSpec setting at the start of DoLoadCore,
so we wouldn't have an incorrect arch value, but that broke the
exhaustive search for kernel binaries, because we didn't have an
address word size or endianness.

This patch should navigate the needs of both use cases.

I spent a good bit of time trying to construct a test to capture all
of these requirements -- but it turns out to be a good bit difficult,
encompassing both a genuine kernel corefiles and a microcontroller
firmware corefiles.

rdar://146821929

jimingham commented on lldb/source/Plugins/Process/mach-core/ProcessMachCore.cpp:
```diff
@@ -578,20 +605,47 @@ Status ProcessMachCore::DoLoadCore() {
   CleanupMemoryRegionPermissions();
 
-  ModuleSP exe_module_sp = GetTarget().GetExecutableModule();
+  exe_module_sp = GetTarget().GetExecutableModule();
```
It seems like you've already done exactly this computation at the beginning of LoadBinariesAndSetDYLD.

Why do you need to do it again here?

jimingham replied:
If there's a reason that's then that's all good, but it might be helpful to future us to say why.

JDevlieghere replied:
Plus, if we really need to do it twice, better to move this into a helper method.

jimingham commented:
Beyond the question about the duplicate work, the approach seems reasonable to me.

JDevlieghere reviewed (approved):
LGTM modulo Jim's questions/suggestions.

jasonmolenda committed:
Don't try to set the target arch to the exe module
in the ExhaustiveSearch - we would have done that already
in DoLoadCore if it was possible.  Use the less-preferable
corefile binary's cputype/cpusubtype to seed Target so we
can search.

jasonmolenda committed:
rm sp

jimingham reviewed (approved):
Okay, that makes sense.
